src/compiler/checker.ts

```diff
@@ -338,7 +338,15 @@ module ts {
                 // Locals of a source file are not in scope (because they get merged into the global symbol table)
                 if (location.locals && !isGlobalSourceFile(location)) {
                     if (result = getSymbol(location.locals, name, meaning)) {
-                        break loop;
+                        // Type parameters of a function are in scope in the entire function declaration, including the parameter
+                        // list and return type. However, local types are only in scope in the function body.
+                        if (!(meaning & SymbolFlags.Type) ||
+                            !(result.flags & (SymbolFlags.Type & ~SymbolFlags.TypeParameter)) ||
+                            !isFunctionLike(location) ||
+                            lastLocation === (<FunctionLikeDeclaration>location).body) {

```

Comment:
Why is this necessary? I thought block scope declarations (like local types) are in the `.locals` of the body, and not of the function itself. Shouldn't that fact make this automatically work out? If that is not true, I think it is a better solution than having this special case.

If this condition is necessary, then isn't this broken for let and const as well?